table/orphan_cleanup.go

```diff
@@ -0,0 +1,593 @@
+	}
+
+	result := &OrphanCleanupResult{
+		OrphanFileLocations: orphanFiles,
+		TotalSizeBytes:      totalSize,
+	}
+
+	if cfg.dryRun {
+		return result, nil
+	}
+	deletedFiles, err := deleteFiles(fs, orphanFiles, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to delete orphan files: %w", err)
+	}
+
+	result.DeletedFiles = deletedFiles
+
+	return result, nil
+}
+
+// getReferencedFiles collects all files referenced by current snapshots
+func (t Table) getReferencedFiles(fs iceio.IO) (map[string]bool, error) {
+	referenced := make(map[string]bool)
+	metadata := t.metadata
+
+	for entry := range metadata.PreviousFiles() {
+		referenced[entry.MetadataFile] = true
+	}
+	referenced[t.metadataLocation] = true
+
+	// Add version hint file (for Hadoop-style tables)
+	// Following Java's ReachableFileUtil.versionHintLocation() logic:
+	versionHintPath := filepath.Join(metadata.Location(), "metadata", "version-hint.text")
+	referenced[versionHintPath] = true
+
+	// TODO: Add statistics files support once iceberg-go exposes statisticsFiles()
+
+	for _, snapshot := range metadata.Snapshots() {
+		if snapshot.ManifestList != "" {
+			referenced[snapshot.ManifestList] = true
+		}
+
+		manifestFiles, err := snapshot.Manifests(fs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read manifests for snapshot %d: %w", snapshot.SnapshotID, err)
+		}
+
+		for _, manifest := range manifestFiles {
+			referenced[manifest.FilePath()] = true
+
+			entries, err := manifest.FetchEntries(fs, false)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read manifest entries: %w", err)
+			}
+
+			for _, entry := range entries {
+				referenced[entry.DataFile().FilePath()] = true
+			}
+		}
+	}
+
+	return referenced, nil
+}
+
+func (t Table) scanFiles(fs iceio.IO, location string, cfg *OrphanCleanupConfig) ([]string, int64, error) {
+	var allFiles []string
+	var totalSize int64
+
+	err := walkDirectory(fs, location, func(path string, info stdfs.FileInfo) error {
+		if info.IsDir() {
+			return nil
+		}
+
+		if !info.ModTime().Before(cfg.olderThan) {
+			return nil // Skip files that are newer than or equal to the threshold
+		}
+
+		allFiles = append(allFiles, path)
+		totalSize += info.Size()
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return allFiles, totalSize, nil
+}
+
+func walkDirectory(fsys iceio.IO, root string, fn func(path string, info stdfs.FileInfo) error) error {
+	// Object stores (S3, GCS, etc.) - use ListIO interface for efficient listing
+	if strings.Contains(root, "://") {
+		if listFS, ok := fsys.(iceio.ListIO); ok {
+			return listFS.ListObjects(root, fn)
+		}
+
+		return fmt.Errorf("cannot list directory %s: storage system does not support listing operations", root)
+	}
+
+	// Local filesystem - use traditional recursive traversal
+	return walkLocalFileSystem(fsys, root, fn)
+}
+
+func walkLocalFileSystem(fsys iceio.IO, root string, fn func(path string, info stdfs.FileInfo) error) error {
```

Comment:
this should be hidden behind the interfaces via `LocalFS` instead of needing a separate implementation here